### DISCUSSION TITLE
When reinstalling a package, chocolatey always deletes download cache

### DISCUSSION THREAD

**Discussion by slycordinator** [chocolatey/choco#2698] [Ideas]
When one reinstalls a package, chocolatey:
1) Removes the download cache directory at "C:\Users\$user\AppData\Local\Temp\chocolatey\$package\$version" as part of the uninstall function.
For this, it seems to be calling "remove_cache_for_package(config, packageVersion)" in NugetServices.cs

2) Then, it redownloads the files that had previously been in the cache
3) Then, verifies the checksums of the newly downloaded files

This is not optimal as chocolatey could simply reuse the files in the cache directory, verify their checksums, and redownload only when the sums don't match.

**Comment by pauby** [comment]
This isn't an issue and doesn't use the template, so converting it to a discussion.

**Comment by CMCDragonkai** [comment]
I've found that the cache is never used at all. I'm using `choco` in a CI/CD environment on gitlab. And even after configuring the cache:

```
    - choco config set cacheLocation "${env:CI_PROJECT_DIR}\tmp\choco"
```

It doesn't actually do anything with it, and just downloads new files each time into that location.

**Reply by pauby** [reply]
If you're using it on a CI / CD system, is it not being built from scratch in GitLab each time? What are you expecting to happen?

**Reply by CMCDragonkai** [reply]
What is being built from scratch? I'm just installing node. In the CI/CD logs it says that it redownloading each time, which is quite different from what I expected the cache to do.

**Reply by CMCDragonkai** [reply]
BTW the directory is saved and cached everytime the job runs. So the cache directory is preserved.

**Reply by binki** [reply]
This is the scenario I have. I noticed that one of my builds spends 20 seconds downloading a package. If I manually save the downloaded files, boot up a fresh Windows environment, bootstrap Chocolatey, and restore the `ChocolateyScratch` folder, Chocolatey still wipes those files away and redownloads them. I want to save on that download time in my CI builds.

**Reply by binki** [reply]
Also, restoring `%PROGRAMDATA%\Chocolatey\lib` doesn’t work because Chocolatey treats the presence of files in this directory as a flag that the package is installed.

Chocolatey needs to learn from NuGet which actually supports using checksums to avoid redownloading assets unnecessarily.

**Comment by pauby** [comment]
> What is being built from scratch?

You're CI / CD system. It was a question rather than a statement.

> I'm just installing node. In the CI/CD logs it says that it redownloading each time, which is quite different from what I expected the cache to do.

It will use something that is already in the cache for a specific period of time. I'm not sure what that is off the top of my head but from experience it's not long. Remember, it is a cache, not a repository.

If you want it to download from somewhere local, you'll look at setting up a repository. You could simply add a folder on your CI / CD system with internalized packages there.

**Reply by CMCDragonkai** [reply]
I'm rerunning the jobs within a couple minutes. And I observed that the cache is always not used. Choco just downloads anew all the time.

**Reply by pauby** [reply]
I mentioned previously:

> I'm not sure what that is off the top of my head but from experience it's not long

The cache is used to cache packages that have been downloaded for a very short period of time. You don't have control over that and Chocolatey will decide whether it needs to download a package again or not. 

It sounds like you want Chocolatey to use that rather than redownloading it again. That's not the job of the cache. That's the job of a repository.

**Reply by CMCDragonkai** [reply]
How short are we talking about here? Is 10 minutes the expiry time for the cache?

**Reply by slycordinator** [reply]
> It will use something that is already in the cache for a specific period of time. I'm not sure what that is off the top of my head but from experience it's not long. Remember, it is a cache, not a repository.

But since it defaults to leave the file in the cache indefinitely, that behavior doesn't make sense, to me.

**Reply by pauby** [reply]
Just to reiterate what I said, Chocolatey CLI manages the cache. It determines when it has to download a file again.

You _can't_ and _shouldn't_ rely on _anything_ in the cache.

The cache isn't a long-term file store. It's a _temporary_ file store where Chocolatey CLI can store what it needs to on a _temporary_ basis. The files may be cached in there for a long time, but it's still a temporary file store.

If you need files to be in a known location and Chocolatey to use them instead of downloading installers / binaries from the internet, then you need to create a repository to do that. 

If you are doing a `--force` install, the cache isn't used. When you run a `choco uninstall` the package you are removing is removed from the cache if it's still there.

I've raised [this issue](https://github.com/chocolatey/docs/issues/475) to allow us to dig into the code and document it further. @AdmiringWorm has raised [this issue](https://github.com/chocolatey/choco/issues/2724) to allow us to evaluate adding the ability to reinstall with using `--force` (which bypasses the cache).

Meanwhile, please set up a repository and install from that. Please don't rely on the cache.

> But since it defaults to leave the file in the cache indefinitely, that behavior doesn't make sense, to me.

Agreed. It does add confusion.

**Reply by binki** [reply]
We don’t need Chocolatey to require the file to be in the cache. We just want it to be able to actually recognize that the file from the cache is still relevant to save download bandwidth and time.

**Comment by slycordinator** [comment]
"Chocolatey CLI manages the cache. It determines when it has to download a
file again."

And at times it does so by ignoring the existing file, which is silly.

"If you are doing a --force install, the cache isn't used."

Which again is silly. If I install a program then immediately realize that
I wanted to enable some package option, I can force it to reinstall. And at
that point, it erases the cache, downloads the same files to the cache
again, then verifies the integrity of those cached files, and finally
installs them. This is especially silly since by default, it will keep those cached files
forever.

**Reply by CMCDragonkai** [reply]
Yea I agree, confusion here is a strange word.
